internal/ssdownload/download_manager.go

```diff
@@ -0,0 +1,243 @@
+		dm.workers[i] = newDownloadWorker(dm, i)
+	}
+	dm.recoveryComplete = make(chan struct{})
+	go dm.recoverDownloads(fftypes.Now())
+	return nil
+}
+
+func (dm *downloadManager) Name() string {
+	return "SharedStorageDownloadManager"
+}
+
+func (dm *downloadManager) WaitStop() {
+	dm.cancelFunc()
+	for _, w := range dm.workers {
+		<-w.done
+	}
+}
+
+func (dm *downloadManager) calcDelay(attempts int) time.Duration {
+	delay := dm.retryInitDelay
+	for i := 0; i < attempts; i++ {
+		delay = time.Duration(math.Ceil(float64(delay) * dm.retryFactor))
+	}
+	if delay > dm.retryMaxDelay {
+		delay = dm.retryMaxDelay
+	}
+	return delay
+}
+
+// recoverDownloads grabs all pending operations on startup, to restart them
+func (dm *downloadManager) recoverDownloads(startupTime *fftypes.FFTime) {
+
+	defer close(dm.recoveryComplete)
+	recovered := 0
+	pageSize := uint64(25)
+	page := uint64(0)
+	errorAttempts := 0
+	for {
+		fb := database.OperationQueryFactory.NewFilter(dm.ctx)
+		filter := fb.And(
+			fb.In("type", []driver.Value{
+				fftypes.OpTypeSharedStorageDownloadBatch,
+				fftypes.OpTypeSharedStorageDownloadBlob,
+			}),
+			fb.Eq("status", fftypes.OpStatusPending),
+			fb.Lt("created", startupTime), // retry is handled completely separately
+		).
+			Sort("created").
+			Skip(page * pageSize).
+			Limit(pageSize)
+		pendingOps, _, err := dm.database.GetOperations(dm.ctx, filter)
+		if err != nil {
+			log.L(dm.ctx).Errorf("Error while recovering pending downloads (retries=%d): %s", errorAttempts, err)
+			errorAttempts++
+			time.Sleep(dm.calcDelay(errorAttempts))
+			continue
+		}
+		errorAttempts = 0 // reset on success
+		page++
+		if len(pendingOps) == 0 {
+			log.L(dm.ctx).Infof("Download manager completed startup after recovering %d pending downloads", recovered)
+			return
+		}
+		for _, op := range pendingOps {
+			preparedOp, err := dm.PrepareOperation(dm.ctx, op)
+			if err != nil {
+				log.L(dm.ctx).Errorf("Failed to recover pending download %s/%s: %s", op.Type, op.ID, err)
+				continue
+			}
+			log.L(dm.ctx).Infof("Recovering pending download %s/%s", op.Type, op.ID)
+			dm.dispatchWork(&downloadWork{
+				op:         op,
+				preparedOp: preparedOp,
+			})
+		}
+	}
+
+}
+
+func (dm *downloadManager) dispatchWork(work *downloadWork) {
+	dm.work <- work
+	// Log after dispatching so we can see the dispatch delay if the queue got full
+	log.L(dm.ctx).Debugf("Dispatched download operation %s/%s (attempts=%d) to worker pool", work.preparedOp.Type, work.preparedOp.ID, work.attempts)
+}
+
+// waitAndRetryDownload is a go routine to wait and re-dispatch a retrying download.
+// Note this go routine is short lived and completely separate to the workers.
+func (dm *downloadManager) waitAndRetryDownload(work *downloadWork) {
+	startedWaiting := time.Now()
+	delay := dm.calcDelay(work.attempts)
+	<-time.After(delay)
+	delayTimeMS := time.Since(startedWaiting).Milliseconds()
+	totalTimeMS := time.Since(*work.op.Created.Time()).Milliseconds()
```

Comment:
Do we really want to know when the `op` was created - or when the `work` was added? Seems to be just for debugging, but in the case of a crash and restore, the `op` creation time could be an hour or a day in the past.

Side note, this seems to be the only place `work.op` is used - wondering if we really need to save `op` itself or if we just need to save a timestamp.

Reply:
Yeah, was for debug. I'm sure I had a point where I was using just a timestamp, and found I needed the full op - but agree that's not true now. So moving to a timestamp of when it was added.